Allow ordering as a guest, even after a you made an account

Hlavtox opened the pull request:
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | See below
| Type?             | improvement
| Category?         | CO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | Fixes https://github.com/PrestaShop/PrestaShop/issues/23007, Fixes https://github.com/PrestaShop/PrestaShop/issues/25355
| How to test?      | See below
| Possible impacts? | -

### Changes
- This PR allows to make an order as a guest even if a registered customer customer exists. This fixes a nonsense behavior when there could be a thousand guest accouts before making a customer account, but you were unable to order as a guest afterwards.
- You are now able to freely change guest account emails. It's checked only if the customer is not a guest. Error message was improved to - `You can't update the email to "john.doe@prestashop.com", because a registered customer with this email already exists.` I will add an automatic test in another PR, after https://github.com/PrestaShop/PrestaShop/pull/31375 is merged, because I can't create guests in tests now.
- In webservice, emails are also checked only in case of registered customers.

### Technical
- I have prepared my ground a bit in another PRs, so it's not much changes.
- Webservice
  - classes/Customer.php - Improved webservice validation.
  - classes/form/CustomerForm.php - I removed validation from the form.
- FO
  - classes/form/CustomerFormatter.php - Attempt to disable autocomplete on password register fields, not sure if browsers will follow. They should.
  - classes/form/CustomerPersister.php - Changed condition and transferred the error message from form validation, so it doesn't change.
  - Front controllers - removed some extra parameters from a call - they are default.
- Admin
  - EditCustomerHandler.php - Added the guest condition to email check.
  - CustomerController.php - Improved the error message 

### How to test FO, guest orders enabled
- Order something and register yourself, log out.
- Go order something and enter the same email as before AND a password. ➡️ You cannot continue as normal, error shown.
- Go order something and enter the same email as before, don't enter password. ➡️ You can still continue in the order as a guest. ✅
- Finish the order and see that **it will not show you the box to convert your account**, because registered customer with that email already exists. - Tackled before in https://github.com/PrestaShop/PrestaShop/pull/25395 ✔️

### How to test FO, guest orders disabled
- Disable guest ordering in Prestashop settings.
- Go order something and enter the same email as before.
- See error that this customer already exists.

### How to test BO
- Go to BO >Customers
- Have three customers:
  - krystian@prestashop.com, registered
  - hiba@prestashop.com, registered
  - matthieu@prestashop.com, guest
- See that:
  - You can change matthieu's email to anything.
  - You cannot change krystian's email to hiba or hiba to krystian.

Hlavtox committed:
Allow ordering for guests, improve guest existence check

PierreRambaud commented on controllers/front/OrderConfirmationController.php:
```diff
@@ -97,6 +97,7 @@ public function initContent()
             'HOOK_PAYMENT_RETURN' => $this->displayPaymentReturn($order),
             'order' => $presentedOrder,
             'register_form' => $register_form,
+            'registered_customer_exists' => Customer::customerExists($this->context->customer->email, false, true),
```
```suggestion
            'registered_customer_exists' => Customer::customerExists($this->context->customer->email, false, true),
```

What if it's a guest? the customer will be empty no?

Hlavtox replied:
@PierreRambaud The last true parameters says, that it should search for guests only. Customer exists -> returns true, Nobody or guests exist -> returns false.

PierreRambaud replied:
I think about `$this->context->customer->email` the `$this->context->customer` can be null

Hlavtox replied:
@PierreRambaud Original code is using the variable two lines down, it's removed there. Cannot be empty, in normal case...

Progi1984 commented:
@Hlavtox I added label "Waiting for PM" as the issue #23007 is waiting for @PrestaShop/product-team

ldeprestashop commented:
Hello @Progi1984 why did you add W4PM ? 

How can I help you to unblock this issue ?

Progi1984 commented:
Check the message from @hibatallahAouadni asking if it's a good / wanted improvement or not.

MatShir commented:
@Hlavtox IMO we should warn the user that he already has an account and he can login with it.

Hlavtox commented:
@MatShir That would be awesome, but there is currently no JS going on in the checkout, everything is validated on PHP side. We can do it, if somebody helps me :-P

MatShir commented:
I suggest something more simple like "You already have an account, don't hesitate to log in!"

MatShir commented:
Check on slack if a maintainer can help you ;)

Hlavtox commented:
So we make a hidden `<div class="alert alert-info" role="alert" style="display: none;">This is a info alert—check it out!</div>`.

On email field change -> AJAX to the controller -> `if (Customer::customerExists($customer->email, false, true)) { return true; }` -> if true, show the message. if false, hide it

Oksydan commented:
@Hlavtox done. Sorry I forgot about it completely.

Oksydan commented:
1. If guest checkout is disabled: 
- password is always visible,
- if customerExists we show error message and prevent from submitting form. Allowing to submit at this point will result only with just showing same error. There is no point to allow user submitting point at this moment.
2. If guest checkout is enabled:
- password field showing and hiding depending if customer exists or not. 
- if customer exits we are showing info message `You already have an account, don't hesitate to log in!` - customer is able to submit form.

Hlavtox commented:
@Progi1984 Tests restarted, can you ping for approval and QA?

Progi1984 commented on themes/classic/templates/customer/_partials/customer-form.tpl:
```diff
@@ -27,7 +27,7 @@
   {/block}
 
-<form action="{block name='customer_form_actionurl'}{$action}{/block}" id="customer-form" class="js-customer-form" method="post">
+<form action="{block name='customer_form_actionurl'}{$action}{/block}" id="customer-form" class="{block name='customer_form_classes'}js-customer-form{/block}" method="post">
```
```suggestion
<form action="{block name='customer_form_actionurl'}{$action}{/block}" id="customer-form" class="{block name='customer_form_class'}js-customer-form{/block}" method="post">
```

Progi1984 commented on themes/classic/templates/checkout/_partials/customer-form.tpl:
```diff
@@ -24,14 +24,18 @@
  *}
 {extends "customer/_partials/customer-form.tpl"}
 
+{block name='customer_form_classes'}js-customer-form js-checkout-customer-form{/block}
```
```suggestion
{block name='customer_form_class'}js-customer-form js-checkout-customer-form{/block}
```

Progi1984 commented on themes/classic/templates/checkout/_partials/customer-form.tpl:
```diff
@@ -24,14 +24,18 @@
+
 {block "form_field"}
   {if $field.name === 'password' and $guest_allowed}
+    <div class="js-checkout-only-new-customer" style="display:none;">
```
```suggestion
    <div class="js-checkout-only-new-customer d-none">
```

Progi1984 commented on themes/_core/js/checkout-customer-form.js:
```diff
@@ -0,0 +1,87 @@
+  let timeout;
+
+  const handleBlockDisplay = ({customerExists}) => {
+    if (customerExists) {
+      $checkoutNewCustomerRelatedBlock.hide();
+    } else {
+      $checkoutNewCustomerRelatedBlock.show();
+    }
```
```suggestion
    $checkoutNewCustomerRelatedBlock.toggleClass('d-none', !customerExists);
```

Progi1984 commented on controllers/front/OrderController.php:
```diff
@@ -237,6 +237,46 @@ public function displayAjaxselectDeliveryOption()
         ]));
     }
 
+    public function displayAjaxcheckCustomerInformation()
```
```suggestion
    public function displayAjaxcheckCustomerInformation(): void
```

Progi1984 commented on controllers/front/OrderController.php:
```diff
@@ -237,6 +237,46 @@ public function displayAjaxselectDeliveryOption()
+    {
+        $email = Tools::getValue('email');
+        $isEmailValid = Validate::isEmail($email);
+        $isGuestAllowed = Configuration::get('PS_GUEST_CHECKOUT_ENABLED');
```
```suggestion
        $isGuestAllowed = (bool) Configuration::get('PS_GUEST_CHECKOUT_ENABLED');
```

Progi1984 commented on controllers/front/OrderController.php:
```diff
@@ -237,6 +237,46 @@ public function displayAjaxselectDeliveryOption()
+        if ($customerExists) {
+            $alert['type'] = $isGuestAllowed ? 'info' : 'danger';
+            $alert['message'] = $isGuestAllowed ?
+            $this->translator->trans(
```
```suggestion
                $this->translator->trans(
```

Progi1984 reviewed (commented):
Some feedback

Oksydan commented:
Thanks for your feedback @Progi1984.

Julievrz commented on controllers/front/OrderController.php:
```diff
@@ -237,6 +237,46 @@ public function displayAjaxselectDeliveryOption()
+            $alert['type'] = $isGuestAllowed ? 'info' : 'danger';
+            $alert['message'] = $isGuestAllowed ?
+                $this->translator->trans(
+                    'You already have an account, don\'t hesitate to log in!',
```
```suggestion
                    'You already have an account, don\'t hesitate to sign in!',
```

Please use "sign in" for more consistency. :-)
 
![image](https://user-images.githubusercontent.com/70583503/128192981-86c0e607-226a-48e5-80b3-98c2f182a877.png)

Oksydan commented:
Hi @Julievrz,
it is done.

Julievrz commented:
> Hi @Julievrz,
> it is done.

Thank you @Oksydan, I have just approved the wording for this PR. :)

Hlavtox commented:
@Julievrz Thank you! Can you remove waiting for author? 🙏 

@Oksydan Thank you!

matthieu-rolland commented on themes/_core/js/checkout-customer-form.js:
```diff
@@ -0,0 +1,83 @@
+export default function () {
+  const $body = $('body');
+  const {checkoutNewCustomerRelatedBlock} = prestashop.selectors.checkout;
+  const checkoutLoginAlertClass = 'js-checkout-login-alert';
```
is there a reason for not putting this class in `themes/_core/js/selectors.js` and call it from there, as you did for `checkoutLoginAlertClass` ?

matthieu-rolland replied:
(is it because of the dot ?)

Oksydan replied:
@matthieu-rolland yes, since it is not only selector but also class for alert block (alert is created dynamically via js). I could just place it inside `selectors.js` files and then remove dot but it might create problem in the future if someone will make nested selector for example.

matthieu-rolland reviewed (commented):
small feedback, lgtm otherwise

Hlavtox commented:
@khouloudbelguith I think it's caused by autofilling the password field, and it gets sent even though it's hidden.

@Oksydan Can we empty the password input when handleBlockDisplay is shown? And ideally, prevent it's autocomplete?

kpodemski commented on controllers/front/OrderController.php:
```diff
@@ -237,6 +237,46 @@ public function displayAjaxselectDeliveryOption()
+        $responseData = [];
+        $alert = [];
+
+        if ($isEmailValid) {
```
```suggestion
        if (Validate::isEmail($email)) {
```

kpodemski commented on controllers/front/OrderController.php:
```diff
@@ -237,6 +237,46 @@ public function displayAjaxselectDeliveryOption()
+    public function displayAjaxcheckCustomerInformation(): void
+    {
+        $email = Tools::getValue('email');
+        $isEmailValid = Validate::isEmail($email);
```
```suggestion
```

kpodemski commented on controllers/front/OrderController.php:
```diff
@@ -237,6 +237,46 @@ public function displayAjaxselectDeliveryOption()
         ]));
     }
 
+    public function displayAjaxcheckCustomerInformation(): void
```
```suggestion
    public function displayAjaxCheckCustomerInformation(): void
```

Hlavtox commented:
Hi @khouloudbelguith, can you please try again? :-) Now it should work fine, the field won't be autocompleted and even if you manually fill it before entering email, it will empty it.

PierreRambaud reviewed (dismissed):
I approve but can you @NeOMakinG have a check. I'm not sure about js in front. Should we really store these files in themes/_core directory?

Hlavtox commented:
Rebased and merged into 3 commits for clarity

NeOMakinG reviewed (dismissed):
Front part looks fine to me

Progi1984 commented on controllers/front/OrderController.php:
```diff
@@ -237,6 +237,45 @@ public function displayAjaxselectDeliveryOption()
+
+        ob_end_clean();
+        header('Content-Type: application/json');
+        $this->ajaxRender(Tools::jsonEncode($responseData));
```
```suggestion
        $this->ajaxRender(json_encode($responseData));
```

Progi1984 replied:
@Hlavtox It seems that this modification has not been applied.

Progi1984 reviewed (commented):
@Hlavtox Thank you for your contribution. I have asked a change because the `Tools::jsonEncode` is deprecated.

Progi1984 reviewed (commented):
@Hlavtox A last modification

Hlavtox commented:
@Progi1984 Sorry, I think I forgot to fetch it before rebasing. All should be good now 👍

PierreRambaud commented on controllers/front/OrderController.php:
```diff
@@ -237,6 +237,45 @@ public function displayAjaxselectDeliveryOption()
+            'customerExists' => $customerExists,
+        ]);
+
+        ob_end_clean();
```
I don't get it :sweat_smile:

PierreRambaud commented on controllers/front/OrderController.php:
```diff
@@ -237,6 +237,45 @@ public function displayAjaxselectDeliveryOption()
+                );
+        }
+
+        $responseData = array_merge($responseData, [
```
```suggestion
        $responseData = [
```

$responseData is never filled.

Progi1984 reviewed (commented):
Excepted feedback from @PierreRambaud, I'm ok with the code.

PierreRambaud reviewed (changes requested):
Why use ob_end_clean, there is no reason to clear the output buffer here?

Hlavtox commented:
@PierreRambaud Removed ;-)

khouloudbelguith commented:
Hi @Hlavtox,

I have two issues:

**First issue**
1. Create a new customer
2. Go to BO > Customers > Customers page
3. Delete the last customer created by selecting this option `I do not want my customer(s) to register again with the same email address. All selected customer(s) will be removed from this list but their corresponding data will be kept in the database.`
4. Go to FO > Add a product to the cart > Proceed to checkout
5. On the checkout page > 1 PERSONAL INFORMATION > Set the same email and click continue (the new late is well displayed) => redirection to the `2 ADDRESSES` step
6. At this step > click on `edit 1 PERSONAL INFORMATION` 
![image](https://user-images.githubusercontent.com/16067358/141776242-94a8e251-487e-448b-8852-f79b4f7ae4d5.png)
7. click on sign in
8. set the address mail + password and click on continue => redirection to the Addresses page => **NOK**
9. If you click again on `edit 1 PERSONAL INFORMATION`  => see the alert  
![image](https://user-images.githubusercontent.com/16067358/141776566-32be1eeb-2e40-487d-abe7-a457eea29680.png)


**Second issue**
1. Create a new customer
2. Go to BO > Customers > Customers page
3. Go to FO > Add a product to the cart > Proceed to checkout
4. On the checkout page > 1 PERSONAL INFORMATION > Set the same email and click continue (the new late is well displayed) => redirection to the `2 ADDRESSES` step
5. At this step > click on `edit 1 PERSONAL INFORMATION` 
![image](https://user-images.githubusercontent.com/16067358/141776242-94a8e251-487e-448b-8852-f79b4f7ae4d5.png)
6. If you edit those fields: Social title, check/uncheck `Receive offers from our partners` & `Sign up for our newsletter
You may unsubscribe at any moment. For that purpose, please find our contact info in the legal notice.` => the alert `You already have an account, don't hesitate to sign in!` is not displayed.

I attached a screen record
https://watch.screencastify.com/v/rSBsmSsahKUVGHuJLW2J

Thanks!

PierreRambaud commented:
I don't know why the waiting for dev was added, but @Hlavtox could you have a look please?

Hlavtox commented:
@PierreRambaud I must check Khouloud's feedback properly and see what must be improved. ;-)

Hlavtox commented:
Rebased, will check and fix the remaining issues today. 👍

Hlavtox commented:
Hi @khouloudbelguith, we just tested the code throughoutly with Valentin.

The first issue is unrelated to this PR. It happens also on develop branch and it produces a weird behavior afterwards, when you are "logged in" (but really not) as the deleted customer it seems.

The second issue we tested throughoutly and we didn't encouter it. When we click back on "edit" button of the first step, the blue message pops up back in 300ms delay as it should.

Can you maybe re-test with @hibatallahAouadni? Really crucial to rebuild core.js assets in this PR. Thank you! :-)

khouloudbelguith reviewed (dismissed):
Hi @Hlavtox,

I found a difference between with & without PR:

1. Disable guest checkout 
2. Go to FO > Add a product to cart & Proceed to checkout 
3. In the information step, add an old address mail 
4. See red the alert `An account was already registered with this email address` 
![image](https://user-images.githubusercontent.com/16067358/154940356-aad35bfa-ea77-4f13-9d84-c224833084db.png)

**Without PR** => I have this red alert `The email is already used, please choose another one or sign in` 
![image](https://user-images.githubusercontent.com/16067358/154940541-15d331c2-aaad-4519-9a75-65878c6c7014.png)

-------------------------------------------------------------------------
I have an issue following these steps:
1. Guest checkout enabled
2. Create an account `demo@prestashop.com`
3. Sign out & add a product to the cart and proceed to checkout
4. In the FO > Checkout page "Personal information": Set all required fields & a new address mail `demo_new@prestashop.com`
5. Before clicking on continue > Modify the address mail to `demo@prestashop.com` & Click enter with the keyboard
6. Set the birthdate
7. See error 
![image](https://user-images.githubusercontent.com/16067358/154941202-295d9e39-73e6-4222-bbcc-cdaa0a5fda6e.png)

I attached a screen record

https://user-images.githubusercontent.com/16067358/154941269-1e6c8049-aa44-4b7b-b26d-1706932d9939.mp4


-------------------------------------------------------------------------
I have an issue following these steps:

1. Guest checkout disabled
2. Create an account `demo@prestashop.com`
3. Sign out & add a product to the cart and proceed to checkout
4. In the FO > Checkout page "Personal information": Set all required fields & a new address mail `demo_new@prestashop.com`
5. Before clicking on continue > Modify the address mail to `demo@prestashop.com` & Click enter with the keyboard
6. Set the password again
7. See an error 
![image](https://user-images.githubusercontent.com/16067358/154941856-67b1874c-a9a1-4814-814f-51dfc986ee4a.png)

I attached a screen record

https://user-images.githubusercontent.com/16067358/154941935-f2e5eb58-4a2b-41ff-a641-4fad4c824ead.mp4

-----------------------------------------------------------
The automatic tests are NOK with this PR also
[guest.pdf](https://github.com/PrestaShop/PrestaShop/files/8108598/guest.pdf)

Thanks!

atomiix commented on classes/form/CustomerPersister.php:
```diff
@@ -33,104 +33,128 @@ class CustomerPersisterCore
-                $clearTextPassword,
-                _COOKIE_KEY_
-            );
+        if ($ok) {
+            if ($guest_to_customer_conversion) {
+                $customer->cleanGroups();
+                $customer->addGroups([Configuration::get('PS_CUSTOMER_GROUP')]);
+                $this->sendConfirmationMail($customer);
+            }
+            $this->context->updateCustomer($customer);
+            $this->context->cart->update();
+            Hook::exec('actionCustomerAccountUpdate', [
+                'customer' => $customer,
+            ]);
```
```suggestion
        if ($customer->save()) {
            if ($guest_to_customer_conversion) {
                $customer->cleanGroups();
                $customer->addGroups([Configuration::get('PS_CUSTOMER_GROUP')]);
                $this->sendConfirmationMail($customer);
            }
            $this->context->updateCustomer($customer);
            $this->context->cart->update();
            Hook::exec('actionCustomerAccountUpdate', [
                'customer' => $customer,
            ]);
            
            return true;
```

atomiix commented on classes/form/CustomerPersister.php:
```diff
@@ -33,104 +33,128 @@ class CustomerPersisterCore
-                    [],
-                    'Shop.Notifications.Error'
-                );
+        return $ok;
```
```suggestion
        return false;
```

atomiix commented on classes/form/CustomerPersister.php:
```diff
@@ -33,104 +33,128 @@ class CustomerPersisterCore
         }
 
-        $guest_to_customer = false;
+        $ok = $customer->save();
```
```suggestion
```

PierreRambaud reviewed (changes requested):
Unfortunately, cs fixer is not really happy :sweat_smile:

Hlavtox commented:
bump

Progi1984 reviewed (dismissed):
Needs a rebase

Hlavtox commented:
QA note - installing related PR is needed for this to work.

khouloudbelguith reviewed (commented):
Hi @Hlavtox,

I found a difference between with & without PR:

1. Disable guest checkout 
2. Go to FO > Add a product to cart & Proceed to checkout 
3. In the information step, add an old address mail 
4. See red the alert `An account was already registered with this email address` 
![image](https://user-images.githubusercontent.com/16067358/154940356-aad35bfa-ea77-4f13-9d84-c224833084db.png)

**Without PR** => I have this red alert `The email is already used, please choose another one or sign in` 
![image](https://user-images.githubusercontent.com/16067358/154940541-15d331c2-aaad-4519-9a75-65878c6c7014.png)

-------------------------------------------------------------------------

1. Enable guest checkout
2. Got to FO > Add a product to the cart & proceed to checkout 
3. In the Information step > the password field is not displayed 
![image](https://user-images.githubusercontent.com/16067358/163590033-c0bf6746-dc62-4337-87b5-b6b3511d59a3.png)

**Without PR**
![image](https://user-images.githubusercontent.com/16067358/163590144-b05236d2-2b40-4264-acf4-e04f79b1eb24.png)


same problem following these steps:
1. Enable guest checkout
2. Got to FO > Add a product to the cart & proceed to checkout 
3. In the Information step > set required fields (without password) and click on continue
4. Edit the Personal information to fill password (to convert the guest to a customer)
5. See error: the password is not displayed  
![image](https://user-images.githubusercontent.com/16067358/163590614-a2a7ce40-6ba1-4556-8d28-6cdc48bfa282.png)

---------------------------------------------------------------------

I have an issue following these steps:
1. Guest checkout disabled
2. Create an account `demo@prestashop.com`
3. Sign out & add a product to the cart and proceed to checkout
4. In the FO > Checkout page "Personal information": Set all required fields & a new address mail `demo_new@prestashop.com` & fill the password
5. Before clicking on continue > Modify the address mail to `demo@prestashop.com` & Click enter with the keyboard
7. See error: the alert is not correct: "Password is incorrect", it should be "an account already registered"

I attached a screen record

https://user-images.githubusercontent.com/16067358/163591120-957793dc-a1ae-4b46-9bb3-0233b1afff64.mp4

--------------------------------------------------------------

I have an issue following these steps:
1. Guest checkout disabled
2. Create an account `demo@prestashop.com`
3. Sign out & add a product to the cart and proceed to checkout
4. In the FO > Checkout page > Personal information step set the old mail `demo@prestashop.com`
5. Try to fill the password
6. See error: the password is emptied automatically

This is normal behavior or a bug?

I attached a screen record

https://user-images.githubusercontent.com/16067358/163591490-d1f4bb35-013e-4465-b2cc-19ec870fcd4e.mp4

--------------------------------------------------

Finnaly, automatic tests are NOK
[guest.pdf](https://github.com/PrestaShop/PrestaShop/files/8496794/guest.pdf)

Thanks!

kpodemski commented:
Hi @Hlavtox @Oksydan 

What about this PR?

kpodemski commented:
Hi @Hlavtox @Oksydan 

If you want to see this PR in PrestaShop 8.1, it's the last call to have it done. The end of January starts the feature freeze for this version.

Oksydan commented:
If you need any help @Hlavtox with js part feel free to DM via slack.

Hlavtox commented:
@kpodemski @matthieu-rolland @marsaldev Guys, I removed the JS part, rebased it, improved it, rewritten the description. It's ready to review.

No BC breaks, no template modifications needed.

Hlavtox commented:
@kpodemski Can you approve and send it to QA?

sallemiines commented:
Hello @Hlavtox 
Thanx for the pr , it LGTM QA :heavy_check_mark: 

I checked  the :

**How to test FO, guest orders enabled :heavy_check_mark:** 

https://user-images.githubusercontent.com/31663369/219378269-bfcee41a-9972-41ee-8d10-662e96023729.mp4


 **How to test FO, guest orders disabled  :heavy_check_mark:**
 
https://user-images.githubusercontent.com/31663369/219379505-95579c5f-d832-40e4-a37e-0ddfc3f92edc.mp4

**How to test BO  :heavy_check_mark:**


https://user-images.githubusercontent.com/31663369/219387801-18222b9a-4eee-4e31-a9e4-f7a3a91714f2.mp4



Thank you :rocket:

kociokwik commented:
Hello @Hlavtox 

I'm merging your changes into 1.7.8.8 (also [#25395](https://github.com/PrestaShop/PrestaShop/pull/25395)) and I found two problems - I'm not sure if they exist in current version of the codebase, but I had to make some changes.

> Main problem with duplicated Customers (duplicated email address, is_guest = 0) is you can't log into one that has higher ID in DB, because of how login works.

1. In OrderConfirmationController::postProcess() (same thing in GuestTrackingController) there's no check if customer with the same email address already exists so it's possible to transform guest into customer with duplicated email address. This means you can't log into newly created account, cause password will be checked against original customer. I had to add this check in the beginning to prevent duplicates:

```php
//added
if ($this->customer->is_guest == 1 && Customer::customerExists($this->customer->email)) {
    $this->errors[] = $this->trans(
        'An account was already registered with this email address',
        [],
        'Shop.Notifications.Error'
    );
} elseif ($this->customer->is_guest == 0) {
    $this->errors[] = $this->trans(
        'A customer account has already been created from this guest account. Please sign in.',
        [],
        'Shop.Notifications.Error'
    );
// Attempt to convert the customer
} elseif ($this->customer->transformToCustomer($this->context->language->id, $password)) {
    $this->success[] = $this->trans(
        'Your guest account has been successfully transformed into a customer account. You can now log in as a registered shopper.',
        [],
        'Shop.Notifications.Success'
    );
}
```

2. In CustomerForm you removed email check, which allowed guest to place an order even if there's already a Customer with this email address, but this enables changing email address in MyAccount for one that is already used. Again, this creates situation where you can't sign into account. To prevent this, I've added this code in CustomerPersister::update()

```php
if (!$customer->is_guest) {
    $customer->passwd = $this->crypto->hash(
        $newPlainTextPassword ? $newPlainTextPassword : $plainTextPassword,
        _COOKIE_KEY_
    );

    //added
    if ($customer->email != $this->context->customer->email && Customer::customerExists($customer->email)) {
        $this->errors['email'][] = $this->translator->trans(
            'An account was already registered with this email address',
            [],
            'Shop.Notifications.Error'
        );

        return false;
    }
}
```

Hlavtox commented:
@kociokwik I also fixed this. ;-)

Check all the PRs I made regarding this topic - https://github.com/PrestaShop/PrestaShop/pulls?q=is%3Apr+conversion+is%3Aclosed+author%3AHlavtox

kociokwik commented:
@Hlavtox 
Regarding first situation I was able to make 2 orders with the same email address as guest, and while OrderConfirmation was opened (for first order) I could go to GuestTracking on another tab (for the second order) and create two customers with the same email, because check (Customer::customerExists($customer->email)) is only done when rendering the page, not for actual transformation request. Not sure if it's a real life problem, but at least it's a possibility to create a dead account

For the second problem I was able to just go into MyAccount and change customer's email address to one that's already used by another customer, because check is done only when transforming guest to customer (while in my scenario only email changes)

I merged pretty much everything regarding this topic besides BO js/twig stuff, cause we just don't use it.

So just one question - does develop branch use src/Adapter/Customer/CommandHandler/EditCustomerHandler.php for requests in FO? If not, I have no idea how is this fixed, but at least it works for me now

Hlavtox commented:
@kociokwik Hmmmm. Please, could you try to reproduce the two bugs on a clean 8.1.6 install? If it's not fixed, please create an issue and ping me, we can fix it together.

> So just one question - does develop branch use src/Adapter/Customer/CommandHandler/EditCustomerHandler.php for requests in FO? If not, I have no idea how is this fixed, but at least it works for me now

It doesn't, command and query bus, along with most symfony functions are not available in FO, because there is only a small container with limited amount of services.

kociokwik commented:
@Hlavtox I have unmodified 1.7.8.8 running currently in production and just installed 8.1.6:
1. 
  * 1.7.8.8 - There's no bug, cause OrderConfirmationController::postProcess() isn't implemented
  * 8.1.6 and develop - You can create two customers with the same email, postProcess doesn't check if customer with provided email exists
2.   
  * 1.7.8.8 and 8.1.6 - you can't reproduce issue because CustomerForm::validate() checks the email at the start.
  * Develop branch - issue exists, cause check was removed from CustomerForm::validate()

I've created an [issue](https://github.com/PrestaShop/PrestaShop/issues/36421) for the first bug, and for the second one my solution from [here](https://github.com/PrestaShop/PrestaShop/pull/25353#issuecomment-2186840234) works
